File: sdks/r/R/meta.R
#' Per-request metadata from the nRouter response headers
#'
#' Every field is \code{NULL} when the gateway did not send it. The gateway
#' omits a header rather than sending a placeholder, and the two omissions that
#' matter most are \code{x-nr-request-cost} — ABSENT when the model is unpriced,
#' never \code{0} — and \code{x-nr-limit-source}, absent when nothing measured a
#' refusal.
#'
#' @param headers Named character vector or list of response headers, keyed by
#'   lowercase name.
#' @return An object of class \code{nrouter_meta}.
#' @export
nrouter_meta <- function(headers = list()) {
  names(headers) <- tolower(names(headers))
  get_chr <- function(name) {
    value <- headers[[name]]
    if (is.null(value) || !nzchar(as.character(value))) NULL else as.character(value)
  }
  # suppressWarnings: a non-numeric header yields NA, which becomes NULL. A
  # zero here would be indistinguishable from a real zero.
  get_num <- function(name) {
    raw <- get_chr(name)
    if (is.null(raw)) return(NULL)
    value <- suppressWarnings(as.numeric(raw))
    if (is.na(value)) NULL else value
  }
  # WHOLE, UNSIGNED DIGITS ONLY, and the regex is the point: `as.numeric` alone
  # accepts "4.5" and `as.integer` silently TRUNCATES it to 4, so a mangled
  # header becomes a plausible-looking measurement. Every other SDK is
  # integer-strict here (Go ParseUint, Rust parse::<u64>, Dart int.tryParse,
  # Kotlin toLongOrNull, Swift Int.init, Java Long.valueOf, JS /^[0-9]+$/), and
  # this is what keeps R from being the one that disagrees.
  get_int <- function(name) {
    raw <- get_chr(name)
    if (is.null(raw)) return(NULL)
    if (!grepl("^[0-9]+$", raw)) return(NULL)
    value <- suppressWarnings(as.numeric(raw))
    if (is.na(value)) NULL else value
  }

  structure(
    class = "nrouter_meta",
    list(
      request_id         = get_chr("x-nr-request-id"),
      # Time to HEADERS, not to the last byte: on a streamed response the
      # headers are ready before the first token.
      latency_ms         = get_int("x-nr-latency-ms"),
      # The gateway's own trace id. A caller may SEND x-nr-trace-id (and
      # x-nr-session-id); the gateway overwrites the response value with its
      # own, so join on this rather than on what was sent.
      trace_id           = get_chr("x-nr-trace-id"),
      cost               = get_num("x-nr-request-cost"),
      cost_status        = get_chr("x-nr-cost-status"),
      model              = get_chr("x-nr-model"),
      input_tokens       = get_num("x-nr-input-tokens"),
      output_tokens      = get_num("x-nr-output-tokens"),
      total_tokens       = get_num("x-nr-total-tokens"),
      cache_read_tokens  = get_num("x-nr-cache-read-tokens"),
      cache_write_tokens = get_num("x-nr-cache-write-tokens"),
      limit_source       = get_chr("x-nr-limit-source"),
      # Posture only: `none`|`monitor`|`pass`|`partial`|`blocked`, matched
      # exactly and case-sensitively. NULL is "no guardrail claim made",
      # never "no guardrail applied" — that is the explicit "none".
      guardrails         = get_chr("x-nr-guardrails"),
      budget_warning     = get_chr("x-nr-budget-warning"),
      auth_reason        = get_chr("x-nr-auth-reason"),
      response_cache     = get_chr("x-nr-response-cache"),
      response_cache_age = get_num("x-nr-response-cache-age"),
      funding_source     = get_chr("x-nr-funding-source"),
      allowance_reset    = get_int("x-nr-allowance-reset"),
      retry_after        = nrouter_parse_retry_after(get_chr("retry-after"))
    )
  )
}

#' Every response header this SDK reads
#'
#' Exactly the names in \code{spec/nrouter-sdk-spec.json}.
#' @return A character vector of the header names, in the order the gateway
#'   emits them.
#' @export
nrouter_header_names <- function() {
  c(
    "x-nr-request-id",
    "x-nr-latency-ms",
    "x-nr-trace-id",
    "x-nr-request-cost",
    "x-nr-cost-status",
    "x-nr-model",
    "x-nr-input-tokens",
    "x-nr-output-tokens",
    "x-nr-total-tokens",
    "x-nr-cache-read-tokens",
    "x-nr-cache-write-tokens",
    "x-nr-limit-source",
    "x-nr-budget-warning",
    "x-nr-guardrails",
    "x-nr-auth-reason",
    "x-nr-response-cache",
    "x-nr-response-cache-age",
    "x-nr-funding-source",
    "x-nr-allowance-reset"
  )
}

nrouter_is_priced <- function(meta) {
  identical(meta$cost_status, "exact") && !is.null(meta$cost)
}

#' Parse structured budget warning from response metadata
#'
#' @param meta An \code{nrouter_meta} object or character string.
#' @return A list with scope, spend, ceiling, or NULL if absent/unparseable.
#' @export
nrouter_parse_budget_warning <- function(meta) {
  raw <- if (inherits(meta, "nrouter_meta")) meta$budget_warning else meta
  if (is.null(raw) || !nzchar(as.character(raw))) return(NULL)
  raw <- trimws(as.character(raw))
  parts <- strsplit(raw, "\\s+")[[1]]
  if (length(parts) != 3 || parts[2] != "soft_budget") return(NULL)
  scope <- parts[1]
  amounts <- strsplit(parts[3], "/")[[1]]
  if (length(amounts) != 2) return(NULL)
  spend <- suppressWarnings(as.numeric(amounts[1]))
  ceiling <- suppressWarnings(as.numeric(amounts[2]))
  if (is.na(spend) || is.na(ceiling) || spend < 0 || ceiling <= 0) return(NULL)
  list(scope = scope, spend = spend, ceiling = ceiling)
}

#' Check if response was a cache hit
#'
#' @param meta An \code{nrouter_meta} object.
#' @return Logical TRUE if cache hit.
#' @export
nrouter_is_cache_hit <- function(meta) {
  identical(meta$response_cache, "hit")
}

#' Check if response was a cache miss
#'
#' @param meta An \code{nrouter_meta} object.
#' @return Logical TRUE if cache miss.
#' @export
nrouter_is_cache_miss <- function(meta) {
  identical(meta$response_cache, "miss")
}

#' Cache age in seconds
#'
#' @param meta An \code{nrouter_meta} object.
#' @return Numeric age in seconds, or 0.
#' @export
nrouter_cache_age_seconds <- function(meta) {
  if (!is.null(meta$response_cache_age)) meta$response_cache_age else 0
}

#' Extract trace routing headers from metadata or a headers list
#'
#' @param source An \code{nrouter_meta} object, a response list, or a named list/vector of headers.
#' @return A named character vector of trace headers.
#' @export
nrouter_extract_trace_headers <- function(source) {
  out <- character(0)
  if (is.null(source)) return(out)
  if (inherits(source, "nrouter_meta")) {
    if (!is.null(source$request_id) && nzchar(as.character(source$request_id))) {
      out["x-nr-request-id"] <- as.character(source$request_id)
    }
    return(out)
  }
  if (is.list(source) && !is.null(source$meta) && inherits(source$meta, "nrouter_meta")) {
    return(nrouter_extract_trace_headers(source$meta))
  }
  headers <- as.list(source)
  names(headers) <- tolower(names(headers))
  for (h in c("x-nr-request-id", "x-nr-trace-id", "x-nr-session-id")) {
    if (!is.null(headers[[h]]) && nzchar(as.character(headers[[h]]))) {
      out[h] <- as.character(headers[[h]])
    }
  }
  out
}

#' Inject trace and session context into an existing headers list or vector
#'
#' Rejects CRLF characters to prevent header injection.
#'
#' @param headers Named list or character vector of headers, or NULL.
#' @param trace_id Optional trace identifier string.
#' @param session_id Optional session identifier string.
#' @return A named list of headers with trace context injected.
#' @export
nrouter_with_trace_context <- function(headers = list(), trace_id = NULL, session_id = NULL) {
  if (!is.null(trace_id) && grepl("[\r\n]", as.character(trace_id))) {
    stop(nrouter_configuration_condition("trace_id must not contain CRLF characters"))
  }
  if (!is.null(session_id) && grepl("[\r\n]", as.character(session_id))) {
    stop(nrouter_configuration_condition("session_id must not contain CRLF characters"))
  }
  out <- list()
  if (!is.null(headers) && length(headers) > 0) {
    nms <- names(headers)
    for (i in seq_along(headers)) {
      v <- as.character(headers[[i]])
      if (!grepl("[\r\n]", v)) {
        out[[nms[i]]] <- v
      }
    }
  }
  if (!is.null(trace_id) && nzchar(as.character(trace_id))) {
    out[["x-nr-trace-id"]] <- as.character(trace_id)
  }
  if (!is.null(session_id) && nzchar(as.character(session_id))) {
    out[["x-nr-session-id"]] <- as.character(session_id)
  }
  out
}


#' @export
print.nrouter_meta <- function(x, ...) {
  cost <- if (is.null(x$cost)) {
    # Unpriced is unknown, not free. Never print it as 0.
    paste0("unpriced", if (!is.null(x$cost_status)) paste0(" (", x$cost_status, ")") else "")
  } else {
    paste0("$", format(x$cost, scientific = FALSE))
  }
  cat("<nrouter_meta>\n")
  cat("  request_id:", x$request_id %||% "-", "\n")
  cat("  model:     ", x$model %||% "-", "\n")
  cat("  cost:      ", cost, "\n")
  invisible(x)
}

`%||%` <- function(a, b) if (is.null(a)) b else a

File: sdks/go/meta.go
package nrouter

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

// ResponseMeta is the per-request metadata the gateway reports on `x-nr-*`
// response headers.
//
// Every numeric field is a POINTER on purpose. The gateway omits a header
// rather than sending a placeholder, and the two that matter most are
// omissions: x-nr-request-cost is ABSENT when the model is unpriced — never
// "0" — and x-nr-response-cache-age is absent on a miss. A float64 zero value
// would render an unpriced request as free, which no enabled model is.
type ResponseMeta struct {
	// RequestID is present on every response; the join key for a spend row or
	// a support ticket.
	RequestID string

	// LatencyMs is the milliseconds the gateway measured from edge arrival to
	// response headers ready.
	//
	// TIME TO HEADERS, not time to the last byte: on a streamed response the
	// headers are ready before the first token, so this is never a
	// total-generation figure. Nil only when the header was absent or
	// unparseable — the edge stamps it on every response it produces.
	LatencyMs *uint64

	// TraceID is the gateway's OpenTelemetry trace id, empty when no valid
	// trace exists.
	//
	// A caller may SEND x-nr-trace-id (and x-nr-session-id) to correlate its
	// own spans; the gateway overwrites the response value with its own, so
	// join on this rather than assuming it echoes what was sent.
	TraceID string

	// Cost is the exact settled cost in USD. Nil when unpriced. Never treat
	// nil as 0.
	Cost *float64

	// CostStatus is "exact" or "unpriced".
	CostStatus string

	// Model is the model that actually served the request, which is not
	// always the alias that was asked for.
	Model string

	InputTokens      *uint64
	OutputTokens     *uint64
	TotalTokens      *uint64
	CacheReadTokens  *uint64
	CacheWriteTokens *uint64

	// LimitSource says which limit measured a 429: key, plan, team, user or
	// budget. Empty means the gateway did not say — never guess, or the
	// customer raises the wrong limit.
	LimitSource string

	// BudgetWarning is set when this request crossed a soft budget you
	// configured; the request still served. Its value is
	// "<scope> soft_budget <spend>/<ceiling>", e.g. "org soft_budget 80.00/100.00".
	BudgetWarning string

	// Guardrails is the posture of the PRE-CALL guardrail chain: "none",
	// "monitor", "pass", "partial" or "blocked". Compare it exactly and
	// case-sensitively.
	//
	// Empty means the gateway made NO guardrail claim about this response — a
	// /v1/models call, an auth refusal that never reached preflight — never "no
	// guardrail applied", which is the explicit "none". Not published on the
	// image, audio or video routes.
	//
	// Posture only, by design: the policy name, its id, the detector family,
	// the rule count and, for "partial", which channel went uninspected are all
	// deliberately withheld. A rule count moves when a policy moves, so a
	// caller watching it maps a tenant's controls without ever tripping one.
	Guardrails string

	// AuthReason is the gateway's stable reason for refusing a virtual key on
	// a 401, e.g. "key_route_not_allowed".
	AuthReason string

	// ResponseCache is "hit" or "miss"; empty when the response cache did not
	// participate at all (streaming and the non-text modalities bypass it).
	ResponseCache string

	// ResponseCacheAge is whole seconds since a cached response was produced.
	// Set on hits only.
	ResponseCacheAge *uint64

	// FundingSource is how this response was funded.
	FundingSource string

	// AllowanceReset is when the current usage allowance resets.
	AllowanceReset *uint64
}

// HeaderNames lists every response header this SDK reads, exactly as the
// gateway spells them. Kept as data so callers can forward the same set
// through their own logging or tracing layer without retyping it.
var HeaderNames = []string{
	"x-nr-request-id",
	"x-nr-latency-ms",
	"x-nr-trace-id",
	"x-nr-request-cost",
	"x-nr-cost-status",
	"x-nr-model",
	"x-nr-input-tokens",
	"x-nr-output-tokens",
	"x-nr-total-tokens",
	"x-nr-cache-read-tokens",
	"x-nr-cache-write-tokens",
	"x-nr-limit-source",
	"x-nr-budget-warning",
	"x-nr-guardrails",
	"x-nr-auth-reason",
	"x-nr-response-cache",
	"x-nr-response-cache-age",
	"x-nr-funding-source",
	"x-nr-allowance-reset",
}

// MetaFromLookup builds ResponseMeta from any lowercase-name header lookup.
//
// An unparseable numeric header yields nil rather than a zero: a zero here
// would be indistinguishable from a real measured zero.
func MetaFromLookup(get func(string) string) ResponseMeta {
	num := func(name string) *uint64 {
		raw := get(name)
		if raw == "" {
			return nil
		}
		// No TrimSpace: MEASURED as dead. net/http strips optional whitespace
		// around a header value before Get returns it, and ParseUint rejects a
		// whitespace-only string, a negative and an out-of-range value on its
		// own. A mutation test proved it — deleting the trim changed no result.
		v, err := strconv.ParseUint(raw, 10, 64)
		if err != nil {
			return nil
		}
		return &v
	}
	meta := ResponseMeta{
		RequestID: get("x-nr-request-id"),
		// num, not a raw read: the gateway sends whole milliseconds, so a
		// fractional or garbage value is a mangled header rather than a
		// latency a caller should chart.
		LatencyMs:        num("x-nr-latency-ms"),
		TraceID:          get("x-nr-trace-id"),
		CostStatus:       get("x-nr-cost-status"),
		Model:            get("x-nr-model"),
		InputTokens:      num("x-nr-input-tokens"),
		OutputTokens:     num("x-nr-output-tokens"),
		TotalTokens:      num("x-nr-total-tokens"),
		CacheReadTokens:  num("x-nr-cache-read-tokens"),
		CacheWriteTokens: num("x-nr-cache-write-tokens"),
		LimitSource:      get("x-nr-limit-source"),
		BudgetWarning:    get("x-nr-budget-warning"),
		Guardrails:       get("x-nr-guardrails"),
		AuthReason:       get("x-nr-auth-reason"),
		ResponseCache:    get("x-nr-response-cache"),
		ResponseCacheAge: num("x-nr-response-cache-age"),
		FundingSource:    get("x-nr-funding-source"),
		AllowanceReset:   num("x-nr-allowance-reset"),
	}
	if raw := get("x-nr-request-cost"); raw != "" {
		if v, err := strconv.ParseFloat(raw, 64); err == nil && isBillableAmount(v) {
			meta.Cost = &v
		}
	}
	return meta
}

// IsPriced reports whether the gateway priced this request exactly. It is
// deliberately not "Cost != nil": a cost with a status of "unpriced" is a
// contradiction the caller should not bill against.
func (m ResponseMeta) IsPriced() bool {
	return m.CostStatus == "exact" && m.Cost != nil
}

// BudgetWarningInfo contains structured soft budget warning details.
type BudgetWarningInfo struct {
	Scope   string
	Spend   float64
	Ceiling float64
}

// ParseBudgetWarning parses BudgetWarning into scope, spend, and ceiling.
func (m ResponseMeta) ParseBudgetWarning() *BudgetWarningInfo {
	if m.BudgetWarning == "" {
		return nil
	}
	var scope string
	var spend, ceiling float64
	n, err := fmt.Sscanf(m.BudgetWarning, "%s soft_budget %f/%f", &scope, &spend, &ceiling)
	if err != nil || n != 3 {
		return nil
	}
	if math.IsNaN(spend) || math.IsInf(spend, 0) || spend < 0 ||
		math.IsNaN(ceiling) || math.IsInf(ceiling, 0) || ceiling <= 0 {
		return nil
	}
	return &BudgetWarningInfo{
		Scope:   scope,
		Spend:   spend,
		Ceiling: ceiling,
	}
}

// IsCacheHit reports whether the response came from the response cache.
func (m ResponseMeta) IsCacheHit() bool {
	return m.ResponseCache == "hit"
}

// IsCacheMiss reports whether the response cache was queried but missed.
func (m ResponseMeta) IsCacheMiss() bool {
	return m.ResponseCache == "miss"
}

// CacheAgeSeconds returns the cached age in seconds, or 0 if absent.
func (m ResponseMeta) CacheAgeSeconds() uint64 {
	if m.ResponseCacheAge != nil {
		return *m.ResponseCacheAge
	}
	return 0
}

// isBillableAmount rejects the values that are syntactically valid floats but
// are not amounts of money.
//
// strconv.ParseFloat accepts "NaN", "Inf", "+Inf" and "Infinity" with a nil
// error, so each one produced a non-nil Cost that a caller would bill against
// — and a NaN then poisons every sum it reaches, silently, forever. A negative
// is rejected for the same reason: the gateway never bills a negative amount,
// so a minus sign means corruption, and a negative quietly netted against a
// bill is worse than a missing one.
//
// A genuine 0 survives. The guard rejects nonsense, not small numbers.
//
// Found by porting this file to TypeScript: Number(”) === 0 forced the same
// question there, and asking it exposed that Go had the wider hole.
func isBillableAmount(v float64) bool {
	return !math.IsNaN(v) && !math.IsInf(v, 0) && v >= 0
}

// ExtractTraceHeaders extracts trace routing headers (e.g. x-nr-request-id) from response metadata.
func ExtractTraceHeaders(meta ResponseMeta) map[string]string {
	out := make(map[string]string)
	if meta.RequestID != "" {
		out["x-nr-request-id"] = meta.RequestID
	}
	return out
}

// WithTraceContext returns a copy of headers with traceID and sessionID injected,
// returning an error if either contains CRLF control characters.
func WithTraceContext(headers map[string]string, traceID, sessionID string) (map[string]string, error) {
	if strings.ContainsAny(traceID, "\r\n") {
		return nil, fmt.Errorf("traceID must not contain CRLF characters")
	}
	if strings.ContainsAny(sessionID, "\r\n") {
		return nil, fmt.Errorf("sessionID must not contain CRLF characters")
	}
	out := make(map[string]string, len(headers)+2)
	for k, v := range headers {
		if !strings.ContainsAny(v, "\r\n") {
			out[k] = v
		}
	}
	if traceID != "" {
		out["x-nr-trace-id"] = traceID
	}
	if sessionID != "" {
		out["x-nr-session-id"] = sessionID
	}
	return out, nil
}
